### PR TITLE
full text search

### PR DESCRIPTION
**Pull request by peillis**

### CODE REVIEW

File: assets/js/search.js [hunk 1]
```diff
@@ -19,100 +19,82 @@ const $input = $('.sidebar-search input')
...
     })
 
     $input.val(value)
-    $search.html(results)
+    $search.html(resultsHtml)
   }
 }
+
+function getIndex () {
+  if (sessionStorage.getItem('idx') == null) {
```
Comment (josevalim) [inline assets/js/search.js]: I believe this can be an issue with Hex. For example, if I am on `hexdocs.pm/foo` and go to `hexdocs.pm/bar` in the same tab, they will share the same tab. Maybe, at the same time we define contentsJSON, we could define a variable called contentsScope, such as "elixir-0.1.0", and use that as key.

File: assets/js/search.js [hunk 2]
```diff
@@ -19,100 +19,82 @@ const $input = $('.sidebar-search input')
...
+    createIndex()
+  }
+  return lunr.Index.load(JSON.parse(sessionStorage.getItem('idx')))
```
Comment (josevalim) [inline assets/js/search.js]: We also need to check what happens if the data is too large to fit the session storage. I assume it will fail to store and getItem will then fail? so it may be better to write this feature in a way that it works even if sessionStorage fails, something like:

```js
var index, stored = sessionStorage.getItem('idx');
if(stored == null) {
  index = createIndex()
  sessionStorage.setItem('idx', JSON.stringify(idx))
} else {
  index = JSON.parse(stored)
}
return lunr.Index.load(index)
```

File: assets/js/search.js (hunk 1 above)
Comment (josevalim) [reply assets/js/search.js]: Alternatively we can store the scope in a `<head><meta>` tag, instead of a JS var.

File: lib/ex_doc/formatter/html.ex [hunk 3]
```diff
@@ -119,6 +120,20 @@ defmodule ExDoc.Formatter.HTML do
     [sidebar_items]
   end
 
+  defp generate_contents_json(nodes_map, extras, config) do
```
Comment (josevalim) [inline lib/ex_doc/formatter/html.ex]: We should probably unify the `sidebar_items` with the `contents_json`. What do you think? But this can also be done in a separate pull request, I thought I would start the discussion though. :)

Comment (peillis) [reply lib/ex_doc/formatter/html.ex]: Right, that makes sense, since I've had to make many things as in ```sidebar_items```

File: assets/js/search.js (hunk 1 above)
Comment (peillis) [reply assets/js/search.js]: Do you mean use the project name as key?

Comment (josevalim) [reply assets/js/search.js]: name+version, yeah. This key can either be passed as a JS variable OR be extracted from the meta (but we need to add it to the meta in the first place).

**Comment by peillis** [comment]
Ok, I've done the changes you suggested. By the way, I don't know why tests fail on travis, I run them locally without problems.

**Comment by josevalim** [comment]
Beautiful! I am currently on holidays and I will take another look at it
once I am back!
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by josevalim** [comment]
Amazing work @peillis! Thank you so much!

I have merged this and I have pushed some improvements to master, namely adding a loading state and loading the contentsJSON only in the search page. I think the contentsJSON late loading is important because for Elixir it is more than 1MB. 

Another change I want to do is to remove the dependency on Poison but once that is done, I think we are almost ready to ship a new version. We will just wait for #967 to merged. I am really excited about this release!

**Comment by peillis** [comment]
You are welcome! I'm very proud to help this amazing community.
